analyzer/src/main/kotlin/managers/utils/DependencyGraphBuilder.kt

```diff
@@ -0,0 +1,324 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.analyzer.managers.utils
+
+import org.ossreviewtoolkit.model.DependencyGraph
+import org.ossreviewtoolkit.model.DependencyReference
+import org.ossreviewtoolkit.model.OrtIssue
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.RootDependencyIndex
+
+/**
+ * A class that can construct a [DependencyGraph] from a set of dependencies that provides an efficient storage format
+ * for large dependency sets in many scopes.
+ *
+ * For larger projects the network of transitive dependencies tends to become complex. Single packages can occur many
+ * times in this structure if they are referenced by multiple scopes or are fundamental libraries, on which many other
+ * packages depend. A naive implementation, which simply duplicates the dependency trees for each scope and package
+ * therefore leads to a high consumption of memory. This class addresses this problem by generating an optimized
+ * structure that shares the components of the dependency graph between the scopes as far as possible.
+ *
+ * This builder class provides the _addDependency()_ function, which has to be called for all the direct dependencies
+ * of the different scopes. From these dependencies it constructs a single, optimized graph that is referenced by all
+ * scopes. To reduce the amount of memory required even further, package identifiers are replaced by numeric indices,
+ * so that references in the graph are just numbers.
+ *
+ * Ideally, the resulting dependency graph contains each dependency exactly once. There are, however, cases, in which
+ * packages occur multiple times in the project's dependency graph with different dependencies. Such cases are
+ * detected, and the corresponding packages form different nodes in the graph, so that they can be distinguished
+ * correctly.
+ *
+ * This class implements the full logic to construct a [DependencyGraph], independent on the concrete representation of
+ * dependencies used by specific package managers. To make this class compatible with such a dependency representation,
+ * the package manager implementation has to provide a [DependencyHandler]. Via this handler, all the relevant
+ * information about dependencies can be extracted.
+ */
+class DependencyGraphBuilder<D>(
+    /**
+     * The [DependencyHandler] used by this builder instance to extract information from the dependency objects when
+     * constructing the [DependencyGraph].
+     */
+    val dependencyHandler: DependencyHandler<D>
+) {
+    companion object {
+        /**
+         * Internal class to represent the result of a search in the dependency graph. The outcome of the search
+         * determines how to integrate a specific dependency into the dependency graph.
+         */
+        private sealed class GraphSearchResult {
```

Comment:
Could this be a private top-level class to have less nesting?

Reply:
Done. At this phase of the rework, there is a clash with the (still existing) private `GraphSearchResult` class in `GradleDependencyGraphBuilder.kt`, and this was also the reason originally why I placed this class in the companion object. To solve this, I renamed it to `DependencyGraphSearchResult`, which I even find a better name.